test/protocol/data/redis/check_redis.c

```diff
@@ -0,0 +1,590 @@
+
+    struct element el_c, el_p;
+    int ret;
+    int len = sizeof(SERIALIZED) - 1;
+    char *pos;
+
+    test_reset();
+
+    /* compose */
+    el_c.type = ELEM_STR;
+    el_c.bstr = str2bstr(STR);
+    ret = compose_element(&buf, &el_c);
+    ck_assert_msg(ret == len, "bytes expected: %d, returned: %d", len, ret);
+    ck_assert_int_eq(cc_bcmp(buf->rpos, SERIALIZED, ret), 0);
+
+    /* parse */
+    pos = buf->rpos + 1;
+    ret = parse_element(&el_p, buf);
+    ck_assert_int_eq(ret, PARSE_OK);
+    ck_assert(buf->rpos == buf->wpos);
+    ck_assert(el_p.type == ELEM_STR);
+    ck_assert(el_p.bstr.len == sizeof(STR) - 1);
+    ck_assert(el_p.bstr.data == pos);
+
+#undef SERIALIZED
+#undef STR
+}
+END_TEST
+
+START_TEST(test_error)
+{
+#define ERR "something is wrong"
+#define SERIALIZED "-" ERR "\r\n"
+
+    struct element el_c, el_p;
+    int ret;
+    int len = sizeof(SERIALIZED) - 1;
+    char *pos;
+
+    test_reset();
+
+    /* compose */
+    el_c.type = ELEM_ERR;
+    el_c.bstr = str2bstr(ERR);
+    ret = compose_element(&buf, &el_c);
+    ck_assert_msg(ret == len, "bytes expected: %d, returned: %d", len, ret);
+    ck_assert_int_eq(cc_bcmp(buf->rpos, SERIALIZED, ret), 0);
+
+    /* parse */
+    pos = buf->rpos + 1;
+    ret = parse_element(&el_p, buf);
+    ck_assert_int_eq(ret, PARSE_OK);
+    ck_assert(buf->rpos == buf->wpos);
+    ck_assert(el_p.type == ELEM_ERR);
+    ck_assert(el_p.bstr.len == sizeof(ERR) - 1);
+    ck_assert(el_p.bstr.data == pos);
+
+#undef SERIALIZED
+#undef ERR
+}
+END_TEST
+
+START_TEST(test_integer)
+{
+#define SERIALIZED_1 ":-1\r\n"
+#define SERIALIZED_2 ":9223372036854775807\r\n"
+#define SERIALIZED_3 ":128\r\n"
+#define INT_1 -1
+#define INT_2 9223372036854775807LL
+#define INT_3 128
+#define OVERSIZE ":19223372036854775807\r\n"
+#define INVALID ":123lOl456\r\n"
+
+    struct element el_c, el_p;
+    int ret;
+
+    test_reset();
+    el_c.type = ELEM_INT;
+
+    el_c.num = INT_1;
+    ret = compose_element(&buf, &el_c);
+    ck_assert(ret == sizeof(SERIALIZED_1) - 1);
+    ck_assert_int_eq(cc_bcmp(buf->rpos, SERIALIZED_1, ret), 0);
+    ret = parse_element(&el_p, buf);
+    ck_assert_int_eq(ret, PARSE_OK);
+    ck_assert(buf->rpos == buf->wpos);
+    ck_assert(el_p.type == ELEM_INT);
+    ck_assert(el_p.num == INT_1);
+
+    buf_reset(buf);
+    el_c.num = INT_2;
+    ret = compose_element(&buf, &el_c);
+    ck_assert(ret == sizeof(SERIALIZED_2) - 1);
+    ck_assert_int_eq(cc_bcmp(buf->rpos, SERIALIZED_2, ret), 0);
+    ret = parse_element(&el_p, buf);
+    ck_assert_int_eq(ret, PARSE_OK);
+    ck_assert(el_p.type == ELEM_INT);
+    ck_assert(el_p.num == INT_2);
+
+    buf_reset(buf);
+    el_c.num = INT_3;
+    ret = compose_element(&buf, &el_c);
+    ck_assert(ret == sizeof(SERIALIZED_3) - 1);
+    ck_assert_int_eq(cc_bcmp(buf->rpos, SERIALIZED_3, ret), 0);
+    ret = parse_element(&el_p, buf);
+    ck_assert_int_eq(ret, PARSE_OK);
+    ck_assert(el_p.num == INT_3);
```

Comment:
missing `ck_assert(el_p.type == ELEM_INT);`

these three could use a local macro to avoid duplication?

```c
#define TEST_SERIALIZE(serialized, value)\
    buf_reset(buf);\
    el_c.num = value;\
    ret = compose_element(&buf, &el_c);\
    ck_assert(ret == sizeof(serialized) - 1);\
    ck_assert_int_eq(cc_bcmp(buf->rpos, serialized, ret), 0);\
    ret = parse_element(&el_p, buf);\
    ck_assert_int_eq(ret, PARSE_OK);\
    ck_assert(el_p.type == ELEM_INT);\
    ck_assert(el_p.num == value);
TEST_SERIALIZE(INT_1, SERIALIZED_1);
TEST_SERIALIZE(INT_2, SERIALIZED_2);
TEST_SERIALIZE(INT_3, SERIALIZED_3);
#undef TEST_SERIALIZE
```